sklearn/utils/isfinite.pyx

```diff
@@ -0,0 +1,152 @@
+# Author: jakirkham
+
+cimport libc
+cimport libc.math
+from libc.math cimport isfinite as _isfinite, isinf as _isinf
+from libc.stdint cimport uint16_t
+
+cimport cython
+
+cimport numpy
+import numpy
+numpy.import_array()
+
+
+cdef extern from "numpy/halffloat.h":
+    ctypedef uint16_t npy_half
+
+    # conversion function
+    float npy_half_to_float(npy_half h);
+
+
+cdef fused fprecision:
+    npy_half
+    float
+    double
+    float complex
+    double complex
+
+cdef fused const_fprecision:
+    const npy_half
+    const float
+    const double
+    const float complex
+    const double complex
+
+
+cdef enum bint_enum:
+    false = False
+    true = True
+
+
+cdef struct bint_false_type:
+    char empty
+
+cdef struct bint_true_type:
+    char empty
+
+cdef fused bint_type:
+    bint_false_type
+    bint_true_type
+
+
+@cython.cdivision(True)
+def py_isfinite(numpy.ndarray a, bint allow_nan=False):
+    cdef numpy.ndarray a_flat
+
+    cdef char* a_data
+    cdef numpy.NPY_TYPES a_type
+    cdef size_t a_step
+    cdef size_t a_size
+
+    cdef bint disallow_nan
+
+    cdef bint result
+    cdef bint_enum err
+
+
+    a_flat = numpy.PyArray_Reshape(a, -1)
```

Comment:
[np.reshape](https://numpy.org/doc/stable/reference/generated/numpy.reshape.html) can sometimes make a copy, which would change the memory usage of this function.

Alternatively, we can restrict the optimization further and only support C or F contiguous arrays.

Reply:
Stipulated the input as C-contiguous.